core/state_processor.go

```diff
@@ -1169,17 +1198,56 @@ func (p *StateProcessor) Process(block *types.WorkObject, batch ethdb.Batch) (ty
 		// Once the total entropy is calculated, the block reward is split
 		// between the blocks, uncles and workshares proportional to the block
 		// weight
-		blockRewardAtTargetBlock := misc.CalculateQuaiReward(targetBlock.Difficulty(), exchangeRate)
+		blockRewardAtTargetBlock := misc.CalculateQuaiReward(targetBlock.WorkObjectHeader(), targetBlock.Difficulty(), exchangeRate)
 		// add the fee capacitor value
 		blockRewardAtTargetBlock = new(big.Int).Add(blockRewardAtTargetBlock, targetBlock.AvgTxFees())
 		// add half the fees generated in the block
 		blockRewardAtTargetBlock = new(big.Int).Add(blockRewardAtTargetBlock, new(big.Int).Div(targetBlock.TotalFees(), common.Big2))
 
+		rewardPerShare := new(big.Int).Div(blockRewardAtTargetBlock, big.NewInt(int64(params.ExpectedWorksharesPerBlock)))
+
 		// Add an etx for each workshare for it to be rewarded
 		for i, share := range sharesAtTargetBlockDepth {
 
-			shareReward := new(big.Int).Mul(blockRewardAtTargetBlock, entropyOfSharesAtTargetBlockDepth[i])
-			shareReward = new(big.Int).Div(shareReward, totalEntropy)
+			var shareReward *big.Int
+			if block.PrimeTerminusNumber().Uint64() < params.KawPowForkBlock {
+				shareReward = new(big.Int).Mul(blockRewardAtTargetBlock, entropyOfSharesAtTargetBlockDepth[i])
+				shareReward = new(big.Int).Div(shareReward, totalEntropy)
+				if shareReward.Cmp(blockRewardAtTargetBlock) > 0 {
+					return nil, nil, nil, nil, 0, 0, 0, nil, nil, errors.New("share reward cannot be greater than the total block reward")
+				}
+			} else {
+
+				shareReward = new(big.Int).Set(rewardPerShare)
+
+				if share.AuxPow() != nil {
+					switch share.AuxPow().PowID() {
+					case types.SHA_BCH, types.SHA_BTC:
+						shareReward = new(big.Int).Mul(shareReward, big.NewInt(int64(totalShaShares)))
+						shareReward = new(big.Int).Div(shareReward, big.NewInt(int64(validShaShares)))
+					case types.Scrypt:
+						shareReward = new(big.Int).Mul(shareReward, big.NewInt(int64(totalScryptShares)))
+						shareReward = new(big.Int).Div(shareReward, big.NewInt(int64(validScryptShares)))
+					}
+				}
+
+				// If mining progpow after the fork, 20% is deducted from the
+				// expectation
+				if share.AuxPow() == nil {
```

Comment:
All these should be made a constant in parameters.

Reply:
made penalty parameters